File: src/p1008/python/solution.py
from typing import List
import math


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def bstFromPreorder(self, v: List[int]) -> TreeNode:
        self.i = 0
        return self.build(v, math.inf)

    def build(self, v, bound):
        if self.i < len(v) and v[self.i] <= bound:
            node = TreeNode(v[self.i])
            self.i += 1
            node.left = self.build(v, node.val)
            node.right = self.build(v, bound)
            return node
        return None

File: src/p1008/golang/solution.go
package main

import (
	"math"
)

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

var i int

func bstFromPreorder(v []int) *TreeNode {
	i = 0
	return build(v, math.MaxInt32)
}

func build(v []int, bound int) *TreeNode {
	if i < len(v) && v[i] <= bound {
		node := TreeNode{Val: v[i]}
		i++
		node.Left = build(v, node.Val)
		node.Right = build(v, bound)
		return &node
	}
	return nil
}
